src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,852 @@
+TypeNameListToObjectAddresses(List *objects)
+{
+	ListCell *objectCell = NULL;
+	List *result = NIL;
+	foreach(objectCell, objects)
+	{
+		TypeName *typeName = castNode(TypeName, lfirst(objectCell));
+		Oid typeOid = LookupTypeNameOid(NULL, typeName, false);
+		ObjectAddress *typeAddress = palloc0(sizeof(ObjectAddress));
+		ObjectAddressSet(*typeAddress, TypeRelationId, typeOid);
+		result = lappend(result, typeAddress);
+	}
+	return result;
+}
+
+
+/*
+ * get_typowner
+ *
+ *		Given the type OID, find its owner
+ */
+static Oid
+get_typowner(Oid typid)
+{
+	Oid result = InvalidOid;
+	HeapTuple tp = NULL;
+
+	tp = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typid));
+	if (HeapTupleIsValid(tp))
+	{
+		Form_pg_type typtup = (Form_pg_type) GETSTRUCT(tp);
+
+		result = typtup->typowner;
+		ReleaseSysCache(tp);
+	}
+
+	return result;
+}
+
+
+/*
+ * makeRangeVarQualified will fill in the schemaname in RangeVar if it is not already
+ * present. The schema used will be the default schemaname for creation of new objects as
+ * returned by RangeVarGetCreationNamespace.
+ */
+static void
+makeRangeVarQualified(RangeVar *var)
+{
+	if (var->schemaname == NULL)
+	{
+		Oid creationSchema = RangeVarGetCreationNamespace(var);
+		var->schemaname = get_namespace_name(creationSchema);
+	}
+}
+
+
+/*
+ * makeTypeNameFromRangeVar creates a TypeName based on a RangeVar.
+ */
+static TypeName *
+makeTypeNameFromRangeVar(const RangeVar *relation)
+{
+	List *names = NIL;
+	if (relation->schemaname)
+	{
+		names = lappend(names, makeString(relation->schemaname));
+	}
+	names = lappend(names, makeString(relation->relname));
+
+	return makeTypeNameFromNameList(names);
+}
+
+
+/*
+ * EnsureSequentialModeForTypeDDL makes sure that the current transaction is already in
+ * sequential mode, or can still safely be put in sequential mode, it errors if that is
+ * not possible. The error contains information for the user to retry the transaction with
+ * sequential mode set from the beginnig.
+ *
+ * As types are node scoped objects there exists only 1 instance of the type used by
+ * potentially multiple shards. To make sure all shards in the transaction can interact
+ * with the type the type needs to be visible on all connections used by the transaction,
+ * meaning we can only use 1 connection per node.
+ */
+static void
+EnsureSequentialModeForTypeDDL(void)
+{
+	if (ParallelQueryExecutedInTransaction())
+	{
+		ereport(ERROR, (errmsg("cannot create or modify type because there was a "
+							   "parallel operation on a distributed table in the "
+							   "transaction"),
+						errdetail("When creating or altering a type, Citus needs to "
+								  "perform all operations over a single connection per "
+								  "node to ensure consistency."),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
+	}
+
+	ereport(DEBUG1, (errmsg("switching to sequential query execution mode"),
```

Comment:
Minor: We should switch to sequential mode if we're inside a transaction block. If we don't do it, we see lots of DEBUG1 messages, which are actually not useful. And, also calling `SetLocalMultiShardModifyModeToSequential` unnecessarily, which is minor again.

Reply:
Guess you mean, switch _only_ when we are in a transaction block?

Thought it was harmless to switch during a single statement, but apparently there are a lot of logs then. Will change